frontend/src/Play.tsx

```diff
@@ -0,0 +1,128 @@
+import * as React from 'react';
+import {IPlayPhase} from './types';
+import Header from './Header';
+import Beeper from './Beeper';
+import Trump from './Trump';
+import Friends from './Friends';
+import Trick from './Trick';
+import Cards from './Cards';
+import Points from './Points';
+import LabeledPlay from './LabeledPlay';
+import Players from './Players';
+import {WebsocketConsumer} from './WebsocketProvider';
+
+type Props = {
+  playPhase: IPlayPhase;
+  name: string;
+  cards: string[];
+  beepOnTurn: boolean;
+  showLastTrick: boolean;
+};
+
+const Play = (props: Props) => {
+  const [selected, setSelected] = React.useState<string[]>([]);
+
+  const playCards = (send: (value: any) => void) => () => {
+    send({Action: {PlayCards: selected}});
+    setSelected([]);
+  };
+
+  const sendEvent = (event: {}) => (send: (value: any) => void) => () => {
+    send(event);
+  };
+  const takeBackCards = sendEvent({Action: 'TakeBackCards'});
+  const endTrick = sendEvent({Action: 'EndTrick'});
+  const startNewGame = sendEvent({Action: 'StartNewGame'});
+
+  const {playPhase} = props;
+
+  // TODO: instead of telling who the player is by checking the name, pass in
+  // the Player object
+  const currentPlayer = playPhase.propagated.players.find(
+    (p) => p.name === props.name,
+  );
+  const nextPlayer = playPhase.trick.player_queue[0];
+  const lastPlay =
+    playPhase.trick.played_cards[playPhase.trick.played_cards.length - 1];
+
+  const isCurrentPlayerTurn = currentPlayer.id === nextPlayer;
+  const canPlay = lastPlay
+    ? selected.length === lastPlay.cards.length
+    : selected.length > 0;
+  const canTakeBack = lastPlay && currentPlayer.id === lastPlay.id;
+
+  const shouldBeBeeping = props.beepOnTurn && isCurrentPlayerTurn;
+
+  const remainingCardsInHands = Object.values(playPhase.hands.hands)
+    .map((playerHand) => {
+      return Object.keys(playerHand).length;
```

Comment:
`playerHand` is a map<Card, usize>, so we need to sum the inner values too right? (although, when zero this is also zero, so it serves its purpose either way)

Reply:
ooooh, I see -- because of the case of duplicate cards. What do you think about using array<Card> format for hands always? I can add some more utils for doing things like add/remove cards from a hand

Will fix in the next commit to this branch

Reply:
the counts data structure is useful on the backend for the actual play validation, and right now the backend datastructure is just serialized to the frontend without modification ( https://github.com/rbtying/shengji/blob/ad156e21d07c369b4912fbe2bb99bd47afa5cd77/core/src/hands.rs#L25 )

I think determining whether the game is over is pretty much the only use case for looking at `hands` right now (they're redacted anyways for other players), since the parallel `cards` array exists to denote ordering

Reply:
fixed in 01110a694de54f995383dbc14c87c313706f86d6